internal/ingress/controller/controller_test.go

```diff
@@ -586,6 +598,338 @@ func TestExtractTLSSecretName(t *testing.T) {
+				}
+			},
+		},
+		{
+			Ingresses: []*ingress.Ingress{
+				{
+					Ingress: extensions.Ingress{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "example",
+							Namespace: "example",
+						},
+						Spec: extensions.IngressSpec{
+							Rules: []extensions.IngressRule{
+								{
+									Host: "example.com",
+									IngressRuleValue: extensions.IngressRuleValue{
+										HTTP: &extensions.HTTPIngressRuleValue{
+											Paths: []extensions.HTTPIngressPath{
+												{
+													Path: "/",
+													Backend: extensions.IngressBackend{
+														ServiceName: "http-svc",
+														ServicePort: intstr.IntOrString{
+															Type:   intstr.Int,
+															IntVal: 80,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					ParsedAnnotations: &annotations.Ingress{
+						Canary: canary.Config{
+							Enabled: false,
+						},
+					},
+				},
+				{
+					Ingress: extensions.Ingress{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "example-canary",
+							Namespace: "example",
+						},
+						Spec: extensions.IngressSpec{
+							Rules: []extensions.IngressRule{
+								{
+									Host: "example.com",
+									IngressRuleValue: extensions.IngressRuleValue{
+										HTTP: &extensions.HTTPIngressRuleValue{
+											Paths: []extensions.HTTPIngressPath{
+												{
+													Path: "/",
+													Backend: extensions.IngressBackend{
+														ServiceName: "http-svc-canary",
+														ServicePort: intstr.IntOrString{
+															Type:   intstr.Int,
+															IntVal: 80,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					ParsedAnnotations: &annotations.Ingress{
+						Canary: canary.Config{
+							Enabled: true,
+						},
+					},
+				},
+			},
+			Validate: func(servers []*ingress.Server) {
+				if len(servers) != 2 {
+					t.Errorf("servers count should be 2, got %d", len(servers))
+					return
+				}
+
+				s := servers[0]
+				if s.Hostname != "_" {
+					t.Errorf("server hostname should be '_', got '%s'", s.Hostname)
+				}
+				if !s.Locations[0].IsDefBackend {
+					t.Errorf("server location 0 should be default backend")
+				}
+
+				if s.Locations[0].Backend != defUpstreamName {
+					t.Errorf("location backend should be '%s', got '%s'", defUpstreamName, s.Locations[0].Backend)
+				}
+
+				s = servers[1]
+				if s.Hostname != "example.com" {
+					t.Errorf("server hostname should be 'example.com', got '%s'", s.Hostname)
+				}
+
+				if s.Locations[0].Backend != "example-http-svc-80" {
+					t.Errorf("location backend should be 'example-http-svc-80', got '%s'", s.Locations[0].Backend)
+				}
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		_, servers := ctl.getBackendServers(testCase.Ingresses)
+		testCase.Validate(servers)
+	}
+
+}
```

Comment:
Loved these tests ❤️ It's also providing a foundation for better unit test coverage in future.